### ISSUE TITLE
Enable secure mode by default

### ISSUE THREAD

**Issue by kaniini** [mastodon/mastodon#18353]
### Pitch

It has been almost 5 years since authorized fetch was first introduced to the fediverse (late 2018).  It seems like it would be reasonable to reevaluate enabling it by default at this point.

### Motivation

Users will benefit in the following ways:

- Increased difficulty for spambots and other unwanted detritus to scrape profile data
- Content will not federate to blocked instances because they will not be able to fetch the content

**Comment by Eriner** [comment]
Threadcap is just one of many would-be casualties of changing this default. A change like this should be made with caution.

**Comment by kaniini** [comment]
This has been implemented since late 2018, in leading implementations.  Unfortunately implementations which have not followed attempts to harden the security model of the fediverse might be inconvenienced but they do not make a compelling argument to avoid moving in the direction of improved security.

**Comment by kaniini** [comment]
I also have to say it’s pretty funny to see somebody so proud of their CVE count to be advocating less security, especially in a context where the legacy AP security model is actively harmful toward anyone wishing to use the security features of AP (e.g. addressing anything other than `as:Public`)

**Comment by Gargron** [comment]
It is worth mentioning that secure mode or lack thereof has zero implications on non-public content in Mastodon, since all non-public content is already behind authentication. Secure mode is specifically for enforcing domain blocks on *public* content. I would say that from an infosec perspective it is a meaningless measure (HTML pages still don't require authentication, so they could be scraped; a dedicated attacker could use a different server to authenticate and then internally mirror/transfer/reuse the data). The measure is meant to stop public content from showing up on blocked servers *casually*, but I would say that it's not worth the loss of performance through caching, the completeness of public threads through forwarding, and the ease of learning and building on the protocol by requiring to have a working implementation before you can even see what live data looks like.

As much as it can be gross to find your public post has made its way onto some iffy server you don't want anything to do with, from an informational perspective it is no different to somebody sharing a link to that post in a Discord you're not aware of, screenshotting it, putting it through archive.is, and so on.

**Comment by Gargron** [comment]
In terms of combatting scraping, it would do a lot to make real-time public timelines require authentication. In 2016/2017 I remember there were a lot of demands to make Mastodon's public timelines publicly accessible so you could look inside a server without signing up or logging in, and tools like unmung.com were created to look inside and so on, but historically, that's exactly the kind of API that's actually been scraped. So that, I would definitely be on board with...

**Comment by Eriner** [comment]
I feel like this setting should be called `REQUIRE_SIGNED_API_READS` or something. "secure mode" is a misnomer and doesn't accurately represent the consequence of enabling it, at least for Mastodon.

Like @Gargon has mentioned most public data is accessible without authentication through the (non-API) rendered HTML web endpoints. It just means that scraping the data requires parsing the HTML, which isn't that big of a deal.

Additionally, if every server in the fediverse were to implement this, it would mean that well-meaning sites/apps/systems looking to allow comments from third-party servers (like Threadcap) now require auth, or must provide and use a default public "dummy" account.

This isn't just a "enable for better security" setting and enabling it by default may not be desirable for many server operators.

**Comment by kaniini** [comment]
> I feel like this setting should be called REQUIRE_SIGNED_API_READS or something. "secure mode" is a misnomer and doesn't accurately represent the consequence of enabling it, at least for Mastodon.

Sure, the setting should be changed to something more granular, and users should be able to opt out of having public profile data.

> Like @Gargon has mentioned most public data is accessible without authentication through the (non-API) rendered HTML web endpoints. It just means that scraping the data requires parsing the HTML, which isn't that big of a deal.

Actually it makes a significant difference in terms of mitigating abuse.  That we can do more is not an argument to do less.

As you’re a security expert you should recognize that security and abuse mitigation at large is a holistic concern.

> Additionally, if every server in the fediverse were to implement this, it would mean that well-meaning sites/apps/systems looking to allow comments from third-party servers (like Threadcap) now require auth, or must provide and use a default public "dummy" account.

This takes at most an hour to implement in an ActivityPub implementation.

> This isn't just a "enable for better security" setting and enabling it by default may not be desirable for many server operators.

Maybe it is undesirable for some, in which case, they disable it in `.env.production` and call it a day.  For most, it is an important control for ensuring that content does not propagate to servers it should not.  

I know that this may come as a surprise to you but there are a number of fediverse servers whose members participate in low effort, high volume harassment, against other fediverse communities they seem “lolcows” and by allowing content to propagate to those servers through unauthorized fetching, encourages drive-by harassment from new instances in that segment of the fediverse which have not yet been blocked.

**Comment by Eriner** [comment]
Your condescension is not productive, nor does it further the argument that the default should be altered.

**Comment by FediVideos** [comment]
Apparently Akkoma (the recent Pleroma fork) has authorized fetch as an option in the admin's graphical interface.

Would it be possible to have this option in Mastodon's admin graphical interface? It would make life a **lot** easier for admins using managed hosting services like masto.host.

![akkoma](https://user-images.githubusercontent.com/75383929/184557451-acf1760b-5bff-4356-834e-04ab1f938eab.png)

**Comment by ShadowJonathan** [comment]
@Eriner 

> Like @Gargon has mentioned most public data is accessible without authentication through the (non-API) rendered HTML web endpoints. It just means that scraping the data requires parsing the HTML, which isn't that big of a deal.

It heightens the barrier, at least, which cuts off half (or more) of the abusers every time you make it incrementally more difficult to engage in a hostile way with these posts.

For example, with secure fetch, and when a user has blocked another, a logged-in user account would be unable to fetch a public post to hit reply to within their web client or app client, as the server would attach their user ID with the request, and the other server will deny serving it on that user.

Ditto with full-server blocks, the server will see that the request is coming from a blocked server, and refuse to serve the programmatic data (JSON) to be able to be acted upon by that server.

Sure, there may be ways around this, ways to alter Mastodon to scrape it from the HTML and reconstruct it themselves (and there are ways to counter *that*), but it would make it incrementally more difficult for bad actors, with low effort, to engage with content this way.

It is not about being exhaustive in the approach, it is about raising the barrier, so that they have no choice but to find/buy a ladder, or give up.

**Comment by kaniini** [comment]
Indeed, it would be nice to see Mastodon 4.0 finally switch the default.  With the large number of new users, many of which are at risk for targeted harassment, having a sensible security mitigation enabled by default would be helpful for mitigating unwanted abuse.

**Comment by ShadowJonathan** [comment]
It should be important to mention that 4.0 now has decoupled authenticated fetch from their rest API, as per https://github.com/mastodon/mastodon/pull/19803

This basically adds more asterisks to authenticated fetch, I hope to see an approach from @Gargron that addresses this properly, allowing secure mode/authenticated fetch to be enabled without problems, and then upgrade to be enabled by default.

The fact that this still hasn't been done does not bring a good case towards gargon, mastodon.social, and other instances about their priorities towards abuse. In my opinion, after having talked with some people about this, there are NO excuses to NOT focus on this, and hide behind problems as excuses as to why this hasn't happened yet.

The developer time was there for a complete redesign of mastodon for v4, there was enough developer time to make a client-side javascript renderer with an individual REST api, there should be enough time left to focus on this, anything less is a priority failure, in my opinion.

**Comment by ThisIsMissEm** [comment]
It's probably worth noting that Authorized Fetch is fairly easily defeated by motivated bad actors: https://soapbox.pub/blog/threads-server-blocking/

So it's debatable whether "Authorized Fetch" is really a "secure mode" since the entire ActivityPub and HTTP Signatures security model is domain based.

**Comment by ShadowJonathan** [comment]
I'd argue so, just that more monitoring metrics and methods are needed to allow someone to track traffic from and to a domain, and notice malicious actors

**Comment by nex3** [comment]
> This takes at most an hour to implement in an ActivityPub implementation.

I would like to push back on this a bit. I've spent _several_ hours today just trying (and failing!) to track down accurate documentation of how "secure mode" even works. I've [documented my trials here](https://nex-3.com/blog/so-im-trying-to-add-support/) if you want the gory details, but the summary is that nothing about this is actually standardized and the standard-adjacent documentation is full of dead links and ambiguities. I genuinely don't think it's possible to implement authorized fetches from scratch without reading and copying an existing implementation.

**Comment by ThisIsMissEm** [comment]
> > This takes at most an hour to implement in an ActivityPub implementation.
> 
> I would like to push back on this a bit. I've spent _several_ hours today just trying (and failing!) to track down accurate documentation of how "secure mode" even works. I've [documented my trials here](https://nex-3.com/blog/so-im-trying-to-add-support/) if you want the gory details, but the summary is that nothing about this is actually standardized and the standard-adjacent documentation is full of dead links and ambiguities. I genuinely don't think it's possible to implement authorized fetches from scratch without reading and copying an existing implementation.

You might find this helpful: https://swicg.github.io/activitypub-http-signature/

It's a draft document outline HTTP Signatures as used today

**Comment by nex3** [comment]
Thank you, that's very helpful!
